poc_mobile_verifier/src/verifier.rs

```diff
@@ -0,0 +1,192 @@
+        let verification_period = reward_period / self.verifications_per_period;
+
+        loop {
+            let now = Utc::now();
+            // Maybe name these "epoch_since_*" and "epoch_since_*_duration"
+            let verify_epoch = self.verifier.get_verify_epoch(now);
+            let reward_epoch = self.verifier.get_reward_epoch(now);
+            let verify_epoch_duration = epoch_duration(&verify_epoch);
+            let reward_epoch_duration = epoch_duration(&reward_epoch);
+
+            // If we started up and the last verification epoch was too recent,
+            // we do not want to re-verify.
+            let mut sleep_duration = if verify_epoch_duration >= verification_period
+                // We always want to verify before a reward 
+                || reward_epoch_duration >= reward_period
+            {
+                tracing::info!("Verifying epoch: {:?}", verify_epoch);
+                // Attempt to verify the current epoch:
+                self.verify_epoch(verify_epoch).await?;
+                verification_period
+            } else {
+                verification_period - verify_epoch_duration
+            };
+
+            // If the current duration since the last reward is exceeded, attempt to
+            // submit rewards
+            if reward_epoch_duration >= reward_period {
+                tracing::info!("Rewarding epoch: {:?}", reward_epoch);
+                self.reward_epoch(reward_epoch).await?
+            } else if reward_epoch_duration + sleep_duration >= reward_period {
+                // If the next epoch is a reward period, cut off sleep duration.
+                // This ensures that verifying will always end up being aligned with
+                // the desired reward period.
+                sleep_duration = reward_period - reward_epoch_duration;
+            }
+
+            tracing::info!("Sleeping...");
+            let shutdown = shutdown.clone();
+            tokio::select! {
+                _ = shutdown => return Ok(()),
+                _ = sleep(
+                    sleep_duration
+                        .to_std()
+                        .map_err(|_| Error::OutOfRangeError)?,
+                ) => (),
+            }
+        }
+    }
+
+    pub async fn verify_epoch(&mut self, epoch: Range<DateTime<Utc>>) -> Result {
+        let transaction = self.pool.begin().await?;
+
+        let shares: Shares = self.verifier.verify_epoch(&epoch).await?;
+
+        // Should we remove the heartbeats that were not new
+        // from valid shares
+        for share in shares.valid_shares.clone() {
+            let heartbeat = Heartbeat::from(share);
+            heartbeat.save(&self.pool).await?;
+        }
+
+        // Update the last verified end time:
+        self.verifier
+            .last_verified_end_time
+            .update(&self.pool, epoch.end.timestamp() as i64)
+            .await?;
+
+        transaction.commit().await?;
```

Comment:
i'm confused about this `transaction` now; what is it even doing? i see you creating and committing it but you're not passing it around anywhere so is it even being used?